packages/cdktf/test/variable.test.ts

```diff
@@ -0,0 +1,107 @@
+import { Testing, TerraformStack, TerraformVariable, TerraformPrimitiveVariableType, TerraformCollectionVariableType, TerraformCollectionType, TerraformTupleVariableType, TerraformObjectVariableType } from '../lib';
+import { TestResource } from './helper'
+
+test('string type', () => {
+    const app = Testing.app();
+    const stack = new TerraformStack(app, 'test');
+
+    new TerraformVariable(stack, 'test-variable', {
+        type: TerraformPrimitiveVariableType.STRING
+    });
+    expect(Testing.synth(stack)).toMatchSnapshot();
+});
+
+test('number type', () => {
+    const app = Testing.app();
+    const stack = new TerraformStack(app, 'test');
+
+    new TerraformVariable(stack, 'test-variable', {
+        type: TerraformPrimitiveVariableType.NUMBER
+    });
+    expect(Testing.synth(stack)).toMatchSnapshot();
+});
+
+test('bool type', () => {
+    const app = Testing.app();
+    const stack = new TerraformStack(app, 'test');
+
+    new TerraformVariable(stack, 'test-variable', {
+        type: TerraformPrimitiveVariableType.BOOL
+    });
+    expect(Testing.synth(stack)).toMatchSnapshot();
+});
+
+test('any type', () => {
+    const app = Testing.app();
+    const stack = new TerraformStack(app, 'test');
+
+    new TerraformVariable(stack, 'test-variable', {
+        type: TerraformPrimitiveVariableType.ANY
+    });
+    expect(Testing.synth(stack)).toMatchSnapshot();
+});
+
+test('default value', () => {
+    const app = Testing.app();
+    const stack = new TerraformStack(app, 'test');
+
+    new TerraformVariable(stack, 'test-variable', {
+        default: 'my-val'
+    });
+    expect(Testing.synth(stack)).toMatchSnapshot();
+});
+
+test('description', () => {
+    const app = Testing.app();
+    const stack = new TerraformStack(app, 'test');
+
+    new TerraformVariable(stack, 'test-variable', {
+        description: 'A Test Variable'
+    });
+    expect(Testing.synth(stack)).toMatchSnapshot();
+});
+
+test('collection type', () => {
+    const app = Testing.app();
+    const stack = new TerraformStack(app, 'test');
+
+    new TerraformVariable(stack, 'test-variable', {
+        type: new TerraformCollectionVariableType(TerraformCollectionType.LIST, TerraformPrimitiveVariableType.STRING)
```

Comment:
That's too much code to define `list(string)` for my taste. I'm wondering if we could simplify this?

Ignoring the `tuple` / `object` type for a moment, this could probably be one enum (e.g. `VariableType.LIST_STRING`, `VariableType.STRING`) - or am I missing something?

- For `tuple`, would this work? `[VariableType.BOOL, VariableType.LIST_NUMBER]
- For `object`, would this work? `{internal: VariableType.NUMBER, protocol: VariableType.STRING]

Then `type` on `TerraformVariable` would have to accept something like `VariableType | VariableType[] | {[key: string]: VariableType | VariableType[] }` (sketched without testing, not sure if that actually works)

Reply:
I'll see if I can cut down on the verbosity, especially for common cases. More complex types may still end up being verbose.

One of the difficulties is that jsii can't handle everything that Typescript can do and resorting to any results in poor experience in other languages especially.

Reply:
> Then `type` on `TerraformVariable` would have to accept something like `VariableType | VariableType[] | {[key: string]: VariableType | VariableType[] }`

I went back and testing union types again and they work better than I remembered.

- Something like that of course works in TS.
- It also seems fine in JS and Python.
- Even Java works pretty well since an overloaded builder method for each type is generated.
- C# is pretty terrible. The result is just an `object`.

Ultimately I want to be writing in C# which is why I went with the implementation that I did; however, I'm not sure that justifies the added verbosity in other languages.

There is hope of [jsii generating more useful union code](https://github.com/aws/aws-cdk-rfcs/blob/feature/remove-type-unions/text/0193-fixing-type-unions.md), but I'm not sure how far out that is. @eladb any more context on that?

Reply:
> I went back and testing union types again and they work better than I remembered.

Thanks for testing this. 

> Ultimately I want to be writing in C# which is why I went with the implementation that I did; however, I'm not sure that justifies the added verbosity in other languages.

Right, I see. Personally, I don't think it'd justify the additional verbosity. In particular when it looks like there's a  [pathway](https://github.com/aws/aws-cdk-rfcs/blob/feature/remove-type-unions/text/0193-fixing-type-unions.md) to a better experience in the future. 

The next step for language support will be Java (which comes more or less for free with #276). C# will have to be added to `jsii-srcmak`.